diff --git a/src/utils/mime.test.ts b/src/utils/mime.test.ts
index 16a8aa41..6d4facde 100644
--- a/src/utils/mime.test.ts
+++ b/src/utils/mime.test.ts
@@ -15,6 +15,7 @@ describe('mime', () => {
     expect(getMimeType('site.webmanifest')).toBe('application/manifest+json')
     expect(getMimeType('goodmorninghellogif')).toBeUndefined()
     expect(getMimeType('indexjs.abcd')).toBeUndefined()
+    expect(getMimeType('IMAGE.JPG')).toBe('image/jpeg')
   })
 
   it('getMimeType with custom mime', () => {
diff --git a/src/utils/mime.ts b/src/utils/mime.ts
index 37e21e11..7db06d11 100644
--- a/src/utils/mime.ts
+++ b/src/utils/mime.ts
@@ -12,7 +12,7 @@ export const getMimeType = (
   if (!match) {
     return
   }
-  let mimeType = mimes[match[1]]
+  let mimeType = mimes[match[1].toLowerCase()]
   if (mimeType && mimeType.startsWith('text')) {
     mimeType += '; charset=utf-8'
   }
