diff --git a/src/middleware/bearer-auth/index.test.ts b/src/middleware/bearer-auth/index.test.ts
index 2b559c02..c9f579c9 100644
--- a/src/middleware/bearer-auth/index.test.ts
+++ b/src/middleware/bearer-auth/index.test.ts
@@ -69,11 +69,80 @@ describe('Bearer Auth by Middleware', () => {
       return c.text('auth-custom-header')
     })
 
+    app.use(
+      '/auth-custom-no-authentication-header-wwwAuthenticateHeader-string/*',
+      bearerAuth({
+        token: tokens,
+        noAuthenticationHeader: {
+          wwwAuthenticateHeader: 'Bearer error="Unauthorized",error_description="Unauthorized"',
+        },
+      })
+    )
+    app.get('/auth-custom-no-authentication-header-wwwAuthenticateHeader-string/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
+    app.use(
+      '/auth-custom-no-authentication-header-wwwAuthenticateHeader-object/*',
+      bearerAuth({
+        token: tokens,
+        noAuthenticationHeader: {
+          wwwAuthenticateHeader: {
+            error: 'Unauthorized',
+            error_description: 'Unauthorized',
+          },
+        },
+      })
+    )
+    app.get('/auth-custom-no-authentication-header-wwwAuthenticateHeader-object/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
+    app.use(
+      '/auth-custom-no-authentication-header-wwwAuthenticateHeader-function-string/*',
+      bearerAuth({
+        token: tokens,
+        noAuthenticationHeader: {
+          wwwAuthenticateHeader: () =>
+            'Bearer error="Unauthorized",error_description="Unauthorized"',
+        },
+      })
+    )
+    app.get(
+      '/auth-custom-no-authentication-header-wwwAuthenticateHeader-function-string/*',
+      (c) => {
+        handlerExecuted = true
+        return c.text('auth')
+      }
+    )
+
+    app.use(
+      '/auth-custom-no-authentication-header-wwwAuthenticateHeader-function-object/*',
+      bearerAuth({
+        token: tokens,
+        noAuthenticationHeader: {
+          wwwAuthenticateHeader: () => ({
+            error: 'Unauthorized',
+            error_description: 'Unauthorized',
+          }),
+        },
+      })
+    )
+    app.get(
+      '/auth-custom-no-authentication-header-wwwAuthenticateHeader-function-object/*',
+      (c) => {
+        handlerExecuted = true
+        return c.text('auth')
+      }
+    )
+
     app.use(
       '/auth-custom-no-authentication-header-message-string/*',
       bearerAuth({
         token,
-        noAuthenticationHeaderMessage: 'Custom no authentication header message as string',
+        noAuthenticationHeader: { message: 'Custom no authentication header message as string' },
       })
     )
     app.get('/auth-custom-no-authentication-header-message-string/*', (c) => {
@@ -85,8 +154,8 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-no-authentication-header-message-object/*',
       bearerAuth({
         token,
-        noAuthenticationHeaderMessage: {
-          message: 'Custom no authentication header message as object',
+        noAuthenticationHeader: {
+          message: { message: 'Custom no authentication header message as object' },
         },
       })
     )
@@ -99,8 +168,9 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-no-authentication-header-message-function-string/*',
       bearerAuth({
         token,
-        noAuthenticationHeaderMessage: () =>
-          'Custom no authentication header message as function string',
+        noAuthenticationHeader: {
+          message: () => 'Custom no authentication header message as function string',
+        },
       })
     )
     app.get('/auth-custom-no-authentication-header-message-function-string/*', (c) => {
@@ -112,9 +182,11 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-no-authentication-header-message-function-object/*',
       bearerAuth({
         token,
-        noAuthenticationHeaderMessage: () => ({
-          message: 'Custom no authentication header message as function object',
-        }),
+        noAuthenticationHeader: {
+          message: () => ({
+            message: 'Custom no authentication header message as function object',
+          }),
+        },
       })
     )
     app.get('/auth-custom-no-authentication-header-message-function-object/*', (c) => {
@@ -122,12 +194,82 @@ describe('Bearer Auth by Middleware', () => {
       return c.text('auth')
     })
 
+    app.use(
+      '/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-string/*',
+      bearerAuth({
+        token: tokens,
+        invalidAuthenticationHeader: {
+          wwwAuthenticateHeader: 'Bearer error="Unauthorized",error_description="Unauthorized"',
+        },
+      })
+    )
+    app.get('/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-string/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
+    app.use(
+      '/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-object/*',
+      bearerAuth({
+        token: tokens,
+        invalidAuthenticationHeader: {
+          wwwAuthenticateHeader: {
+            error: 'Unauthorized',
+            error_description: 'Unauthorized',
+          },
+        },
+      })
+    )
+    app.get('/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-object/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
+    app.use(
+      '/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-function-string/*',
+      bearerAuth({
+        token: tokens,
+        invalidAuthenticationHeader: {
+          wwwAuthenticateHeader: () =>
+            'Bearer error="Unauthorized",error_description="Unauthorized"',
+        },
+      })
+    )
+    app.get(
+      '/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-function-string/*',
+      (c) => {
+        handlerExecuted = true
+        return c.text('auth')
+      }
+    )
+
+    app.use(
+      '/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-function-object/*',
+      bearerAuth({
+        token: tokens,
+        invalidAuthenticationHeader: {
+          wwwAuthenticateHeader: () => ({
+            error: 'Unauthorized',
+            error_description: 'Unauthorized',
+          }),
+        },
+      })
+    )
+    app.get(
+      '/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-function-object/*',
+      (c) => {
+        handlerExecuted = true
+        return c.text('auth')
+      }
+    )
+
     app.use(
       '/auth-custom-invalid-authentication-header-message-string/*',
       bearerAuth({
         token,
-        invalidAuthenticationHeaderMessage:
-          'Custom invalid authentication header message as string',
+        invalidAuthenticationHeader: {
+          message: 'Custom invalid authentication header message as string',
+        },
       })
     )
     app.get('/auth-custom-invalid-authentication-header-message-string/*', (c) => {
@@ -139,8 +281,10 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-invalid-authentication-header-message-object/*',
       bearerAuth({
         token,
-        invalidAuthenticationHeaderMessage: {
-          message: 'Custom invalid authentication header message as object',
+        invalidAuthenticationHeader: {
+          message: {
+            message: 'Custom invalid authentication header message as object',
+          },
         },
       })
     )
@@ -153,8 +297,9 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-invalid-authentication-header-message-function-string/*',
       bearerAuth({
         token,
-        invalidAuthenticationHeaderMessage: () =>
-          'Custom invalid authentication header message as function string',
+        invalidAuthenticationHeader: {
+          message: () => 'Custom invalid authentication header message as function string',
+        },
       })
     )
     app.get('/auth-custom-invalid-authentication-header-message-function-string/*', (c) => {
@@ -166,9 +311,11 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-invalid-authentication-header-message-function-object/*',
       bearerAuth({
         token,
-        invalidAuthenticationHeaderMessage: () => ({
-          message: 'Custom invalid authentication header message as function object',
-        }),
+        invalidAuthenticationHeader: {
+          message: () => ({
+            message: 'Custom invalid authentication header message as function object',
+          }),
+        },
       })
     )
     app.get('/auth-custom-invalid-authentication-header-message-function-object/*', (c) => {
@@ -176,11 +323,74 @@ describe('Bearer Auth by Middleware', () => {
       return c.text('auth')
     })
 
+    app.use(
+      '/auth-custom-invalid-token-wwwAuthenticateHeader-string/*',
+      bearerAuth({
+        token: tokens,
+        invalidToken: {
+          wwwAuthenticateHeader: 'Bearer error="Unauthorized",error_description="Unauthorized"',
+        },
+      })
+    )
+    app.get('/auth-custom-invalid-token-wwwAuthenticateHeader-string/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
+    app.use(
+      '/auth-custom-invalid-token-wwwAuthenticateHeader-object/*',
+      bearerAuth({
+        token: tokens,
+        invalidToken: {
+          wwwAuthenticateHeader: {
+            error: 'Unauthorized',
+            error_description: 'Unauthorized',
+          },
+        },
+      })
+    )
+    app.get('/auth-custom-invalid-token-wwwAuthenticateHeader-object/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
+    app.use(
+      '/auth-custom-invalid-token-wwwAuthenticateHeader-function-string/*',
+      bearerAuth({
+        token: tokens,
+        invalidToken: {
+          wwwAuthenticateHeader: () =>
+            'Bearer error="Unauthorized",error_description="Unauthorized"',
+        },
+      })
+    )
+    app.get('/auth-custom-invalid-token-wwwAuthenticateHeader-function-string/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
+    app.use(
+      '/auth-custom-invalid-token-wwwAuthenticateHeader-function-object/*',
+      bearerAuth({
+        token: tokens,
+        invalidToken: {
+          wwwAuthenticateHeader: () => ({
+            error: 'Unauthorized',
+            error_description: 'Unauthorized',
+          }),
+        },
+      })
+    )
+    app.get('/auth-custom-invalid-token-wwwAuthenticateHeader-function-object/*', (c) => {
+      handlerExecuted = true
+      return c.text('auth')
+    })
+
     app.use(
       '/auth-custom-invalid-token-message-string/*',
       bearerAuth({
         token,
-        invalidTokenMessage: 'Custom invalid token message as string',
+        invalidToken: { message: 'Custom invalid token message as string' },
       })
     )
     app.get('/auth-custom-invalid-token-message-string/*', (c) => {
@@ -192,7 +402,7 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-invalid-token-message-object/*',
       bearerAuth({
         token,
-        invalidTokenMessage: { message: 'Custom invalid token message as object' },
+        invalidToken: { message: { message: 'Custom invalid token message as object' } },
       })
     )
     app.get('/auth-custom-invalid-token-message-object/*', (c) => {
@@ -204,7 +414,7 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-invalid-token-message-function-string/*',
       bearerAuth({
         token,
-        invalidTokenMessage: () => 'Custom invalid token message as function string',
+        invalidToken: { message: () => 'Custom invalid token message as function string' },
       })
     )
     app.get('/auth-custom-invalid-token-message-function-string/*', (c) => {
@@ -216,9 +426,11 @@ describe('Bearer Auth by Middleware', () => {
       '/auth-custom-invalid-token-message-function-object/*',
       bearerAuth({
         token,
-        invalidTokenMessage: () => ({
-          message: 'Custom invalid token message as function object',
-        }),
+        invalidToken: {
+          message: () => ({
+            message: 'Custom invalid token message as function object',
+          }),
+        },
       })
     )
     app.get('/auth-custom-invalid-token-message-function-object/*', (c) => {
@@ -386,6 +598,58 @@ describe('Bearer Auth by Middleware', () => {
     expect(await res.text()).toBe('Unauthorized')
   })
 
+  it('Should not authorize - custom no authorization header wwwAuthenticateHeader as string', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-no-authentication-header-wwwAuthenticateHeader-string'
+    )
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom no authorization header wwwAuthenticateHeader as object', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-no-authentication-header-wwwAuthenticateHeader-object'
+    )
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom no authorization header wwwAuthenticateHeader as function string', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-no-authentication-header-wwwAuthenticateHeader-function-string'
+    )
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom no authorization header wwwAuthenticateHeader as function object', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-no-authentication-header-wwwAuthenticateHeader-function-object'
+    )
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
   it('Should not authorize - custom no authorization header message as string', async () => {
     const req = new Request('http://localhost/auth-custom-no-authentication-header-message-string')
     const res = await app.request(req)
@@ -430,6 +694,62 @@ describe('Bearer Auth by Middleware', () => {
     )
   })
 
+  it('Should not authorize - custom invalid authentication header wwwAuthenticateHeader as string', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-string'
+    )
+    req.headers.set('Authorization', 'Beare abcdefg12345-._~+/=')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(400)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom invalid authentication header wwwAuthenticateHeader as object', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-object'
+    )
+    req.headers.set('Authorization', 'Beare abcdefg12345-._~+/=')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(400)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom invalid authentication header wwwAuthenticateHeader as function string', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-function-string'
+    )
+    req.headers.set('Authorization', 'Beare abcdefg12345-._~+/=')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(400)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom invalid authentication header wwwAuthenticateHeader as function object', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-authentication-header-wwwAuthenticateHeader-function-object'
+    )
+    req.headers.set('Authorization', 'Beare abcdefg12345-._~+/=')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(400)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
   it('Should not authorize - custom invalid authentication header message as string', async () => {
     const req = new Request(
       'http://localhost/auth-custom-invalid-authentication-header-message-string'
@@ -484,6 +804,62 @@ describe('Bearer Auth by Middleware', () => {
     )
   })
 
+  it('Should not authorize - custom invalid token wwwAuthenticateHeader as string', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-token-wwwAuthenticateHeader-string'
+    )
+    req.headers.set('Authorization', 'Bearer invalid-token')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom invalid token wwwAuthenticateHeader as object', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-token-wwwAuthenticateHeader-object'
+    )
+    req.headers.set('Authorization', 'Bearer invalid-token')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom invalid token wwwAuthenticateHeader as function string', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-token-wwwAuthenticateHeader-function-string'
+    )
+    req.headers.set('Authorization', 'Bearer invalid-token')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
+  it('Should not authorize - custom invalid token wwwAuthenticateHeader as function object', async () => {
+    const req = new Request(
+      'http://localhost/auth-custom-invalid-token-wwwAuthenticateHeader-function-object'
+    )
+    req.headers.set('Authorization', 'Bearer invalid-token')
+    const res = await app.request(req)
+    expect(res).not.toBeNull()
+    expect(res.status).toBe(401)
+    expect(handlerExecuted).toBeFalsy()
+    expect(res.headers.get('WWW-Authenticate')).toBe(
+      'Bearer error="Unauthorized",error_description="Unauthorized"'
+    )
+  })
+
   it('Should not authorize - custom invalid token message as string', async () => {
     const req = new Request('http://localhost/auth-custom-invalid-token-message-string')
     req.headers.set('Authorization', 'Bearer invalid-token')
diff --git a/src/middleware/bearer-auth/index.ts b/src/middleware/bearer-auth/index.ts
index 34d8993f..3ecd2e3b 100644
--- a/src/middleware/bearer-auth/index.ts
+++ b/src/middleware/bearer-auth/index.ts
@@ -14,6 +14,10 @@ const PREFIX = 'Bearer'
 const HEADER = 'Authorization'
 
 type MessageFunction = (c: Context) => string | object | Promise<string | object>
+type CustomizedErrorResponseOptions = {
+  wwwAuthenticateHeader?: string | object | MessageFunction
+  message?: string | object | MessageFunction
+}
 
 type BearerAuthOptions =
   | {
@@ -22,9 +26,21 @@ type BearerAuthOptions =
       prefix?: string
       headerName?: string
       hashFunction?: Function
+      /**
+       * @deprecated Use noAuthenticationHeader.message instead
+       */
       noAuthenticationHeaderMessage?: string | object | MessageFunction
+      noAuthenticationHeader?: CustomizedErrorResponseOptions
+      /**
+       * @deprecated Use invalidAuthenticationHeader.message instead
+       */
       invalidAuthenticationHeaderMessage?: string | object | MessageFunction
+      invalidAuthenticationHeader?: CustomizedErrorResponseOptions
+      /**
+       * @deprecated Use invalidToken.message instead
+       */
       invalidTokenMessage?: string | object | MessageFunction
+      invalidToken?: CustomizedErrorResponseOptions
     }
   | {
       realm?: string
@@ -32,9 +48,21 @@ type BearerAuthOptions =
       headerName?: string
       verifyToken: (token: string, c: Context) => boolean | Promise<boolean>
       hashFunction?: Function
+      /**
+       * @deprecated Use noAuthenticationHeader.message instead
+       */
       noAuthenticationHeaderMessage?: string | object | MessageFunction
+      noAuthenticationHeader?: CustomizedErrorResponseOptions
+      /**
+       * @deprecated Use invalidAuthenticationHeader.message instead
+       */
       invalidAuthenticationHeaderMessage?: string | object | MessageFunction
+      invalidAuthenticationHeader?: CustomizedErrorResponseOptions
+      /**
+       * @deprecated Use invalidToken.message instead
+       */
       invalidTokenMessage?: string | object | MessageFunction
+      invalidToken?: CustomizedErrorResponseOptions
     }
 
 /**
@@ -49,9 +77,12 @@ type BearerAuthOptions =
  * @param {string} [options.prefix="Bearer"] - The prefix (or known as `schema`) for the Authorization header value. If set to the empty string, no prefix is expected.
  * @param {string} [options.headerName=Authorization] - The header name.
  * @param {Function} [options.hashFunction] - A function to handle hashing for safe comparison of authentication tokens.
- * @param {string | object | MessageFunction} [options.noAuthenticationHeaderMessage="Unauthorized"] - The no authentication header message.
- * @param {string | object | MessageFunction} [options.invalidAuthenticationHeaderMessage="Bad Request"] - The invalid authentication header message.
- * @param {string | object | MessageFunction} [options.invalidTokenMessage="Unauthorized"] - The invalid token message.
+ * @param {string | object | MessageFunction} [options.noAuthenticationHeader.message="Unauthorized"] - The no authentication header message.
+ * @param {string | object | MessageFunction} [options.noAuthenticationHeader.wwwAuthenticateHeader="Bearer realm=\"\""] - The response header value for the WWW-Authenticate header when no authentication header is provided.
+ * @param {string | object | MessageFunction} [options.invalidAuthenticationHeader.message="Bad Request"] - The invalid authentication header message.
+ * @param {string | object | MessageFunction} [options.invalidAuthenticationHeader.wwwAuthenticateHeader="Bearer error=\"invalid_request\""] - The response header value for the WWW-Authenticate header when authentication header is invalid.
+ * @param {string | object | MessageFunction} [options.invalidToken.message="Unauthorized"] - The invalid token message.
+ * @param {string | object | MessageFunction} [options.invalidToken.wwwAuthenticateHeader="Bearer error=\"invalid_token\""] - The response header value for the WWW-Authenticate header when token is invalid.
  * @returns {MiddlewareHandler} The middleware handler function.
  * @throws {Error} If neither "token" nor "verifyToken" options are provided.
  * @throws {HTTPException} If authentication fails, with 401 status code for missing or invalid token, or 400 status code for invalid request.
@@ -88,11 +119,21 @@ export const bearerAuth = (options: BearerAuthOptions): MiddlewareHandler => {
   const throwHTTPException = async (
     c: Context,
     status: ContentfulStatusCode,
-    wwwAuthenticateHeader: string,
+    wwwAuthenticateHeader: string | object | MessageFunction,
     messageOption: string | object | MessageFunction
   ): Promise<Response> => {
+    const wwwAuthenticateHeaderValue: string | object =
+      typeof wwwAuthenticateHeader === 'function'
+        ? await wwwAuthenticateHeader(c)
+        : wwwAuthenticateHeader
+
     const headers = {
-      'WWW-Authenticate': wwwAuthenticateHeader,
+      'WWW-Authenticate':
+        typeof wwwAuthenticateHeaderValue === 'string'
+          ? wwwAuthenticateHeaderValue
+          : `${wwwAuthenticatePrefix}${Object.entries(wwwAuthenticateHeaderValue)
+              .map(([key, value]) => `${key}="${value}"`)
+              .join(',')}`,
     }
     const responseMessage =
       typeof messageOption === 'function' ? await messageOption(c) : messageOption
@@ -116,8 +157,11 @@ export const bearerAuth = (options: BearerAuthOptions): MiddlewareHandler => {
       await throwHTTPException(
         c,
         401,
-        `${wwwAuthenticatePrefix}realm="${realm}"`,
-        options.noAuthenticationHeaderMessage || 'Unauthorized'
+        options.noAuthenticationHeader?.wwwAuthenticateHeader ||
+          `${wwwAuthenticatePrefix}realm="${realm}"`,
+        options.noAuthenticationHeader?.message ||
+          options.noAuthenticationHeaderMessage ||
+          'Unauthorized'
       )
     } else {
       const match = regexp.exec(headerToken)
@@ -126,8 +170,11 @@ export const bearerAuth = (options: BearerAuthOptions): MiddlewareHandler => {
         await throwHTTPException(
           c,
           400,
-          `${wwwAuthenticatePrefix}error="invalid_request"`,
-          options.invalidAuthenticationHeaderMessage || 'Bad Request'
+          options.invalidAuthenticationHeader?.wwwAuthenticateHeader ||
+            `${wwwAuthenticatePrefix}error="invalid_request"`,
+          options.invalidAuthenticationHeader?.message ||
+            options.invalidAuthenticationHeaderMessage ||
+            'Bad Request'
         )
       } else {
         let equal = false
@@ -148,8 +195,9 @@ export const bearerAuth = (options: BearerAuthOptions): MiddlewareHandler => {
           await throwHTTPException(
             c,
             401,
-            `${wwwAuthenticatePrefix}error="invalid_token"`,
-            options.invalidTokenMessage || 'Unauthorized'
+            options.invalidToken?.wwwAuthenticateHeader ||
+              `${wwwAuthenticatePrefix}error="invalid_token"`,
+            options.invalidToken?.message || options.invalidTokenMessage || 'Unauthorized'
           )
         }
       }
